### PR TITLE
Add schema for V1 contracts

### PR DESCRIPTION
**Pull request by abizjak**
## Purpose

Add support for V1 schemas.

## Changes

Add a return value schema for V1 contracts.

## Checklist

- [ ] My code follows the style of this project.
- [ ] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: src/schema.rs [hunk 1]
```diff
@@ -272,24 +314,41 @@ impl Deserial for Fields {
...
         self.contracts.serial(out)?;
         Ok(())
     }
 }
 
-impl Deserial for Module {
+impl Deserial for ModuleV0 {
     fn deserial<R: Read>(source: &mut R) -> ParseResult<Self> {
         let len: u32 = source.get()?;
         let contracts = deserial_map_no_length_no_order_check(source, len as usize)?;
```
Comment (Bargsteen) [inline src/schema.rs:334]: I think we should just use `deserial`, which does the exact same thing.

File: src/schema.rs [hunk 2]
```diff
@@ -272,24 +314,41 @@ impl Deserial for Fields {
...
             contracts,
         })
     }
 }
 
-impl Serial for Contract {
+impl Deserial for ModuleV1 {
+    fn deserial<R: Read>(source: &mut R) -> ParseResult<Self> {
+        let len: u32 = source.get()?;
+        let contracts = deserial_map_no_length_no_order_check(source, len as usize)?;
```
Comment (Bargsteen) [inline src/schema.rs:344]: And here as well.

File: src/schema.rs [hunk 3]
```diff
@@ -33,26 +33,68 @@ pub trait SchemaType {
...
+#[derive(Debug, Default, Clone)]
+/// The [Default] instance produces an empty schema.
+pub struct ContractV1 {
+    pub init:    Option<FunctionSchema>,
+    pub receive: BTreeMap<String, FunctionSchema>,
+}
+
+/// Describes the schema of an init or a receive function for V1 contracts.
+#[derive(Debug, Clone)]
+pub enum FunctionSchema {
```
Comment (Bargsteen) [inline src/schema.rs:67]: I think this should be named `Function` to be consistent with the rest. We usually use it qualified anyhow, so `schema::Function` is clear on its own.

File: src/schema.rs [hunk 4]
```diff
@@ -298,20 +357,78 @@ impl Serial for Contract {
...
+        Ok(())
+    }
+}
+
+impl Deserial for ContractV0 {
     fn deserial<R: Read>(source: &mut R) -> ParseResult<Self> {
         let state = source.get()?;
         let init = source.get()?;
         let len: u32 = source.get()?;
         let receive = deserial_map_no_length_no_order_check(source, len as usize)?;
```
Comment (Bargsteen) [inline src/schema.rs:373]: And here

File: src/schema.rs [hunk 5]
```diff
@@ -298,20 +357,78 @@ impl Serial for Contract {
...
             receive,
         })
     }
 }
 
+impl Deserial for ContractV1 {
+    fn deserial<R: Read>(source: &mut R) -> ParseResult<Self> {
+        let init = source.get()?;
+        let len: u32 = source.get()?;
+        let receive = deserial_map_no_length_no_order_check(source, len as usize)?;
```
Comment (Bargsteen) [inline src/schema.rs:386]: And here

**Review by Bargsteen — APPROVED** [review]
LGTM, just minor tweaks needed.

File: src/schema.rs (hunk 1 above)
Comment (abizjak) [reply src/schema.rs:334]: I don't particularly want to change this. I think it is OK to be explicit here.

**Pull request merged**